mission/mission_handler_navigating.hxx

```diff
@@ -0,0 +1,207 @@
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f,
+                            0.0f);
+}
+
+template <class T>
+bool Mission_handler_navigating<T>::can_handle(const Waypoint& wpt) const
+{
+    bool handleable = false;
+
+    uint16_t cmd = wpt.command();
+    if (cmd == MAV_CMD_NAV_WAYPOINT)
+    {
+        handleable = true;
+    }
+
+    return handleable;
+}
+
+template <class T>
+bool Mission_handler_navigating<T>::setup(const Waypoint& wpt)
+{
+    bool success = true;
+
+    start_time_ = time_keeper_get_ms();
+    waypoint_reached_ = false;
+    waypoint_ = wpt;
+
+    return success;
+}
+
+template <class T>
+Mission_handler::update_status_t Mission_handler_navigating<T>::update()
+{
+    // Set goal
+    bool ret = set_control_command();
+    
+    /**********************************
+    Determine if arrived for first time 
+    **********************************/
+    // Find distance to waypoint
+    local_position_t wpt_pos = waypoint_.local_pos();
+    float rel_pos[3];
+    for (int i = 0; i < 3; i++)
+    {
+        rel_pos[i] = wpt_pos[i] - ins_.position_lf()[i];
+    }
+    float dist2wp_sqr = vectors_norm_sqr(rel_pos);
+
+    // Check if radius is available
+    float radius;
+    if (!waypoint_.radius(radius))
+    {
+        radius = 4.0f;
+    }
+
+    // If we are near the waypoint or are doing dubin circles
+    if (dist2wp_sqr < (radius * radius))
+    {
+        // If we are near the waypoint but the flag has not been set, do this once ...
+        if (!waypoint_reached_)
+        {
+            // Send debug log ...
+            print_util_dbg_print("Waypoint reached, distance: ");
+            print_util_dbg_putfloat(sqrt(dist2wp_sqr), 3);
+            print_util_dbg_print(" m. Less than acceptable radius:");
+            print_util_dbg_putfloat(sqrt(radius * radius), 3);
+            print_util_dbg_print(" m.\r\n");
+
+            // ... as well as a mavlink message ...
+            mavlink_message_t msg;
+            mavlink_msg_mission_item_reached_pack(mavlink_stream_.sysid(),
+                                                  mavlink_stream_.compid(),
+                                                  &msg,
+                                                  waypoint_handler_.current_waypoint_index());
+            mavlink_stream_.send(&msg);
+
+            // ... and record the travel time ...
+            travel_time_ = time_keeper_get_ms() - start_time_;
+
+            // ... and set to waiting at waypoint ...
+            waypoint_reached_ = true;
+        }
+    }
+
+    /*******************
+    Determine status code 
+    ********************/
+    // First check if we have reached the waypoint
+    if (waypoint_reached_)
+    {
+        // Then ensure that we are in autocontinue
+        if ((waypoint_.autocontinue() == 1) && (waypoint_handler_.waypoint_count() > 1))
```

Comment:
The check for the waypoint count is to not uselessly advance to the same waypoint that you're at. If there's only 1 waypoint and we just arrived at it, should we be advancing to that same waypoint.